### ISSUE TITLE
Failed to deploy helm chart with local dependencies and no internet access

### ISSUE THREAD

**Issue by kuznero** [argoproj/argo-cd#786]
I am trying to deploy `prometheus-operator` from `stable` channel from `github.com/helm/charts`. The chart is pre-pulled with its dependencies like following:

```
helm fetch --untar stable/prometheus-operator
```

The chart is then committed to source control and the Argo CD app created. During deployment it complains about unavailability of internet when trying to download dependencies (there is no internet connection available at all, that is why all dependencies are committed to the repo with `helm fetch --untar`).

```
argocd app get prometheus-operator
Name:               prometheus-operator
Server:             https://kubernetes.default.svc
Namespace:          monitoring
URL:                https://127.0.0.1:8080/applications/prometheus-operator
Repo:               git@stash.danskenet.net:7999/kube/sandbox.git
Target:             HEAD
Path:               charts/prometheus-operator
Helm Values:        values.yaml
Sync Policy:        <none>

CONDITION        MESSAGE
ComparisonError  rpc error: code = Unknown desc = Error: could not download https://kubernetes-charts.storage.googleapis.com/kube-state-metrics-0.11.0.tgz: Get https://kubernetes-charts.storage.googleapis.com/kube-state-metrics-0.11.0.tgz: dial tcp: lookup kubernetes-charts.storage.googleapis.com on 10.96.0.10:53: no such host
```

I have also tried running: `helm dep build` and commit all the archived dependencies. Argo CD still shows the same error trying to download dependencies again.

**Comment by jessesuen** [comment]
I believe the error message "could not download https://kubernetes-charts.storage.googleapis.com/" is coming from when we invoke `helm dependency build` before the `helm template` command. `helm dependency build` was added as part of https://github.com/argoproj/argo-cd/issues/582 (in order to increase Chart support from the main helm repository).

In order to better support firewalled/air-gapped networks, we should only run `helm dep build` if we detect that we need it (e.g. `helm template` failed.

**Comment by kuznero** [comment]
The solution for me was pretty painful - flatten the dependencies into single chart. That way I completely loose chance to pull updates from `helm/charts/stable`.

**Comment by adieu** [comment]
Had the same issue. Waiting for a fix.

**Comment by jessesuen** [comment]
Fixed in v0.11. `helm dependency build` will only be invoked if `helm template` fails with a specific dependency issue, after which the `helm template` command will re-run.
